services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadWithKnownContentLengthHelper.java

```diff
@@ -90,98 +95,173 @@ public CompletableFuture<PutObjectResponse> uploadObject(PutObjectRequest putObj
 
     private void uploadInParts(PutObjectRequest putObjectRequest, long contentLength, AsyncRequestBody asyncRequestBody,
                                CompletableFuture<PutObjectResponse> returnFuture) {
+        String uploadId;
+        HashMap<Integer, CompletedPart> existingParts = null;
+        S3ResumeToken resumeToken = putObjectRequest.overrideConfiguration()
+                                                    .map(c -> c.executionAttributes()
+                                                               .getAttribute(RESUME_TOKEN)).orElse(null);
+
+        if (resumeToken == null) {
+            uploadId = multipartUploadHelper.createMultipartUpload(putObjectRequest, returnFuture).join().uploadId();
```

Comment:
We probably shouldn't block here

Reply:
removed join, reverted to future.whenComplete()